pkg/controllermanager/controller/certificatesigningrequest/certificatesigningrequest.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certificatesigningrequest
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllermanager"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/logger"
+
+	"github.com/prometheus/client_golang/prometheus"
+	kubeinformers "k8s.io/client-go/informers"
+	certificatesv1beta1lister "k8s.io/client-go/listers/certificates/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Controller controls CertificateSigningRequests.
+type Controller struct {
+	k8sGardenClient kubernetes.Interface
+
+	control  ControlInterface
+	recorder record.EventRecorder
+
+	csrLister certificatesv1beta1lister.CertificateSigningRequestLister
+	csrQueue  workqueue.RateLimitingInterface
+	csrSynced cache.InformerSynced
+
+	workerCh               chan int
+	numberOfRunningWorkers int
+}
+
+// NewCSRController takes a Kubernetes client for the Garden clusters <k8sGardenClient>, a struct
+// holding information about the acting Gardener, a <kubeInformerFactory>, and a <recorder> for
+// event recording. It creates a new Gardener controller.
+func NewCSRController(k8sGardenClient kubernetes.Interface, kubeInformerFactory kubeinformers.SharedInformerFactory, recorder record.EventRecorder) *Controller {
```

Comment:
I don’t think we can use existing CSR Auto approver. It hard-codes to the “system:nodes” group, while we have “gardener.cloud:system:seeds”.

I don’t see how this problem can be circumvented by just applying RBAC roles? Am I missing something?

Reply:
> I don’t think we can use existing CSR Auto approver. It hard-codes to the “system:nodes” group, while we have “gardener.cloud:system:seeds”.

No. Existing auto-approver is generic, only the configured `recongnizer` is group specific.  So, in existing `approver` while creating we [configure](https://github.com/kubernetes/kubernetes/blob/release-1.15/pkg/controller/certificates/approver/sarapprove.go#L49) our own implementation of `recognizer` which recognises  `gardener.cloud:system:seeds`. And it will work for us.

> I don’t see how this problem can be circumvented by just applying RBAC roles?

If you are referring to [my this comment](https://github.com/gardener/gardener/pull/1601#discussion_r346254178), Yes it can't just be solved by using RBAC. I should have dropped that comment. 

But yes, using above explained implementation would be preferable to me.

Reply:
I don't understand, sorry. Can you explain in more detail how we can use the existing CSR auto-approver controller?

Reply:
Ahh sorry. Got my mistake. I was assuming [csrRecognizer](https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/controller/certificates/approver/sarapprove.go#L34) and [sarApprovar](https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/controller/certificates/approver/sarapprove.go#L40) structs are exposed. And was thinking something like below. But yes since they are private its not possible. 
```
// Custom-gardenlet-recogniser
func customGardenRecognizers() []csrRecognizer {
	recognizers := []csrRecognizer{
		{
			recognize:      isSeedClientCert,
			permission:     authorization.ResourceAttributes{Group: "certificates.k8s.io", Resource: "certificatesigningrequests", Verb: "create", Subresource: "seedclient"},
			successMessage: "Auto approving gardenlet client certificate after SubjectAccessReview.",
		},
	}
	return recognizers
}
// Customer-gardenlet-approver
// NewCSRApprovingController creates a new CSRApprovingController.
func NewCSRApprovingController(client clientset.Interface, csrInformer certificatesinformers.CertificateSigningRequestInformer) *certificates.CertificateController {
	approver := &sarApprover{
		client:      client,
		recognizers: customGardenRecognizers(),
	}
	return certificates.NewCertificateController(
		client,
		csrInformer,
		approver.handle,
	)
}

```